Screen implementation

JurjenBiewenga opened the pull request.

JurjenBiewenga committed:
Added Win32 screen implementation

JurjenBiewenga committed:
Forgot saving these

JurjenBiewenga committed:
Started work on GTK implementation

JurjenBiewenga committed:
Created a working implementation for Gtk3

JurjenBiewenga committed:
Added logging for testing

JurjenBiewenga committed:
Added gtk implementation

JurjenBiewenga committed:
Added equals, gethashcode, performance improvements

JurjenBiewenga committed:
Resolving conflicts

JurjenBiewenga committed:
Merge branch 'master' into Screen

kekekeks commented on src/Gtk/Avalonia.Gtk3/Interop/Native.cs:
```diff
@@ -20,6 +20,24 @@ static class Native
+            public delegate gint16 gdk_display_get_n_screens(IntPtr display);
+
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl), GtkImport(GtkDll.Gdk)]
+            public delegate GtkScreen gdk_display_get_screen(IntPtr display, gint16 num);
```
According to https://developer.gnome.org/gdk3/stable/GdkDisplay.html#gdk-display-get-screen
```
gdk_display_get_screen has been deprecated since version 3.20 and should not be used in newly-written code.
There is only one screen; use gdk_display_get_default_screen() to get it.
```

kekekeks replied:
Same with `gdk_display_get_n_screens`

kekekeks replied:
You also have to make sure that handle returned by `gdk_display_get_default_screen` is unowned, it will be released by finalizer otherwise.

JurjenBiewenga committed:
Replaced GetScreen with GetDefaultScreen

JurjenBiewenga committed:
Merge

JurjenBiewenga commented:
I wasn't able to find a way to get a callback when the display settings changed on win32 that works in .net core. Any ideas on how to cache the screen data?

JurjenBiewenga committed:
ScreenCount is now cached in gtk3

JurjenBiewenga committed:
Forgot to commit the name changed interface

JurjenBiewenga committed:
Renamed GtkScreen to GdkScreen

JurjenBiewenga committed:
Created UnownedGdkScreen class

JurjenBiewenga committed:
Added constructor to gdkscreen

JurjenBiewenga committed:
Moved Screen data to its own class

JurjenBiewenga committed:
Screen no longer override Equals

JurjenBiewenga committed:
Forcing recompile

JurjenBiewenga commented:
Whoever is interested, feel free to take over the branch. I'm unable to properly develop on it due to not being able to build the project. Until it's fixed I can't do much.

JurjenBiewenga commented:
Reopening PR as the issue has been resolved

JurjenBiewenga committed:
Created platform specific Screen implementations

JurjenBiewenga committed:
Merge

kekekeks commented:
It seems that you have used `git rebase`, so diff is a total mess now (see Files Changed section). Can't really review it like this

JurjenBiewenga committed:
Revert "Merge"

This reverts commit 57b1681b64e049c530f8c52a07f5e0708c4646aa.

JurjenBiewenga committed:
Added files to project

kekekeks commented:
We also have MonoMac backend merged now, so it will require Screen implementation as well. Beware that OSX uses *inverted* Y axis (coordinate system has (0,0) at left *bottom* corner, so we need to do tcoordinate translations when required.

JurjenBiewenga committed:
Merge branch 'master' into Screen

JurjenBiewenga committed:
Added basic mac implementation

JurjenBiewenga committed:
Cached allScreens
Created mac specific implementation of Screen for comparision and hashcodes

JurjenBiewenga committed:
Added caching to win32 implementation

JurjenBiewenga committed:
Changed formatting of privates
Added ConvertRectY to MonoMac.Helpers
MacScreenImpl now uses helper methods to convert to different coordinate systems

JurjenBiewenga committed:
Changed primary screen on osx to the screen with the menu bar

kekekeks commented on src/OSX/Avalonia.MonoMac/Helpers.cs:
```diff
@@ -27,5 +28,14 @@ public static CGPoint ConvertPointY(this CGPoint pt)
 
+        public static Rect ConvertRectY(this Rect rect)
+        {
+            return new Rect(new Point(rect.X, rect.Position.ConvertPointY().Y), rect.Size);
```
This conversion is not correct. In original coordinate system Y is the position of the *bottom* side of the rectangle. In our windows-like coordinate system Y is the position of the *top* side of the rectangle.

kekekeks commented on src/OSX/Avalonia.MonoMac/ScreenImpl.cs:
```diff
@@ -0,0 +1,62 @@
+                    }
+
+                    _allScreens = s;
+                    _observer = NSNotificationCenter.DefaultCenter.AddObserver(NSApplicationDidChangeScreenParametersNotification, notification =>
```
I think this subscription should happen in static constructor instead. Same with GTK implementation, since neither GdkDisplay of GdkScreen would ever change.

kekekeks commented on src/OSX/Avalonia.MonoMac/ScreenImpl.cs:
```diff
@@ -0,0 +1,62 @@
+            }
+        }
+        
+        public Screen PrimaryScreen
```
Since you are going through the array of cached instances, this logic should be moved to `Avalonia.Platform.Screen`

JurjenBiewenga replied:
Whoops, that shouldn't have been the cached instances

JurjenBiewenga committed:
Mac/win32 implementation now use the AllScreens property to find the primary screen

JurjenBiewenga committed:
Moved callback registration to constructor

JurjenBiewenga committed:
Fixed helper functions

JurjenBiewenga committed:
Added Control.Catalog page
Exposed Screens to WindowBase
Fixed bug in win32 implementation where the width would be the top right coordinate

JurjenBiewenga committed:
Moved things around

JurjenBiewenga committed:
Android PopupImpl now has a null Screens variable

JurjenBiewenga committed:
Moved Screens to Window

JurjenBiewenga committed:
Removed remnant of Screens variable

JurjenBiewenga committed:
Added bounds to screen implementation

JurjenBiewenga committed:
Merge branch 'master' into Screen

kekekeks commented on src/Avalonia.Controls/Screens.cs:
```diff
@@ -5,13 +5,19 @@ namespace Avalonia.Controls
             _iScreenImpl = iScreenImpl;
         }
+
+        public Screen ScreenFromBounds(Rect bounds)
```
Why it exists here and in `IScreenImpl`?

kekekeks commented on src/Gtk/Avalonia.Gtk3/ScreenImpl.cs:
```diff
@@ -0,0 +1,71 @@
+        {
+            get
+            {
+                for (int i = 0; i < AllScreens.Length; i++)
```
It should be moved to `Screens.cs`, I think

JurjenBiewenga committed:
Moved a ScreenFromBounds and Primary to Screens class

JurjenBiewenga committed:
Restored draw dirty rects

JurjenBiewenga committed:
Added ScreenFromPoint and ScreenFromVisual

JurjenBiewenga committed:
Merge branch 'master' into Screen

SinZ163 commented on samples/ControlCatalog/MainView.xaml.cs:
```diff
@@ -13,11 +13,20 @@ public MainView()
+            {
+                IList tabItems = ((IList)this.FindControl<TabControl>("Sidebar").Items);
+                tabItems.Add(new TabItem()
+                                                                           {
```
Is this meant to be indented this far?

JurjenBiewenga replied:
This is how Rider does it, it indents to the same column. I prefer it this way but I don't mind changing it back.

SinZ163 commented on samples/ControlCatalog/MainView.xaml.cs:
```diff
@@ -13,11 +13,20 @@ public MainView()
+                                                                               Content = new DialogsPage()
+                                                                           });
+                tabItems.Add(new TabItem()
+                             {
```
This also seems unusually indented

SinZ163 commented on src/Avalonia.Controls/Screens.cs:
```diff
@@ -0,0 +1,57 @@
+        public int ScreenCount => _iScreenImpl.ScreenCount;
+        public Screen[] All => _iScreenImpl?.AllScreens;
+        public Screen Primary
+        {
```
You lambda the previous two properties, why not this one too?

JurjenBiewenga replied:
That used to be a for loop, I'll change it.

SinZ163 commented on src/Gtk/Avalonia.Gtk3/Interop/GObject.cs:
```diff
@@ -31,32 +31,50 @@ protected override bool ReleaseHandle()
+    }
+    
+     class UnownedGdkScreen : GdkScreen
+        {
```
One too many indents?

SinZ163 reviewed (commented):
Various indentation and inconsistent lambas

JurjenBiewenga committed:
Fixed indentation

JurjenBiewenga committed:
Screens.Primary is now consistent with the other properties

JurjenBiewenga committed:
Merge branch 'Screen' of https://github.com/JurjenBiewenga/Avalonia into Screen

JurjenBiewenga committed:
Merge

JurjenBiewenga committed:
Merge branch 'master' into Screen

JurjenBiewenga committed:
Screens is now properly cached

kekekeks commented:
Superseded by https://github.com/AvaloniaUI/Avalonia/commit/e25f949d9889b64748a7c5bf0055413bc0ef115a
